libcaf_core/caf/detail/ring_buffer.hpp

```diff
@@ -0,0 +1,68 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/config.hpp"
+
+#include <vector>
+
+namespace caf::detail {
+
+// A simple ring buffer implementation
+template <class T>
+class ring_buffer {
```

Comment:
A few observations on this container:

- Your implementation will crash when constructing a ring-buffer of size `0` and calling `push_back` on it. None of your tests covers this case.
- Having a `vector` member adds unnecessary overhead. A `std::unique_ptr<T[]>` plus a `size_` member are sufficient for our needs since we don't need re-allocations and capacity management.
- I found the member names a bit confusing. The `curr_back_` member is the writing position, so it's actually the past-the-end position, not the position of the `back` element. I would find `write_pos_` much more intuitive.
- The `curr_front_` member is redundant and I would argue dangerous since it can run out of sync. We can compute this index from the writing position at any time: `auto front_index = (write_pos_ + max_size_ - size_) % max_size_`.
- This implementation monotonically increments the write position, why did you implement it this way instead of `write_pos_ = (write_pos_ + 1) % max_size_`?

I think this container is asking for no more than three `size_t` members (`size_`, `max_size_` and `write_pos_`) plus a `data_` pointer. Your current implementation uses five `size_t` members plus a pointer (`std::vector` has capacity, size and a data pointer).

Reply:
I think that using monotonically increasing write position may give us information related to how much data is written into the buffer if we need to access it. We can remove this easily and the front index can be calculated easily in that scenario.

Reply:
Also, I think there is another issue with the current implementation. There are no checks before pop_front()/front() to handle empty buffers. I think that this should throw an error as we should not return any value from buffer when it is empty.

Reply:
Calling `front` on an empty container is undefined behavior. No need for checks here (the standard library doesn't do checks here either).

Handling a zero-sized buffer gracefully does make sense to stay consistent with the rest of the language (`auto ptr = new int[0]` is valid C++). I think a zero-sized buffer should simply stay `empty`, no matter how often you call `push_back` on it.